### PR TITLE
feat: add ad-hoc task runs

### PR DESCRIPTION
**Pull request by inhumantsar**
*Issue #, if available:* #54 

*Description of changes:* This adds the ability to run tasks outside of a service, either standalone or as an init task. 

This borrows a lot from #273 but offers a slightly cleaner interface and isn't dependent on a service.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

EDIT: In case you're seeing this and want to use it, you can pull it directly from the PR in your GitHub Actions workflow (see below). We've been using it in production for months now and I'm happy to keep supporting it.

```yaml
      - name: Deploy Amazon ECS task definition
        id: deploy-task-def
        uses: inhumantsar/amazon-ecs-deploy-task-definition@feat/run-task
        with:
          task-definition: ${{ steps.render-task-def.outputs.task-definition }}
          ...
```

### CODE REVIEW

**Comment by duggan** [comment]
Thanks @inhumantsar! I've been running this in a fork for a couple of weeks, works reliably, has been very handy for running DB migrations as part of an ECS deployment flow.

Would be great to see it merged to make it generally available.

**Comment by inhumantsar** [comment]
@clareliguori @mattsb42-aws sorry for the ping, just hoping to get a review on this.

**Review by ahmadg-caylent — APPROVED** [review]

**Comment by anitakrueger** [comment]
We would also very much like this feature. We've been using the amazon-ecs-deploy-task-definition action for quite a while and this will enable us to run one time tasks.

**Comment by inhumantsar** [comment]
@duggan @anitakrueger FYI you can use this PR prior to merge. See the updated PR description

**Comment by andrewlynch** [comment]
When a create a once off task, the ENI is being created without a public IP address. It's the same as not ticking the Public IP address box on the AWS Console when running a task. Is it possible to specify that an IP must be created?

**Comment by redterror** [comment]
Would it be possible to adjust this so that the task definition isn't implicitly created / updated?  This feature would help my team but we are managing our task defs / services via terraform and already have an ARN we could pass in.

File: dist/index.js [hunk 1]
```diff
@@ -26,6 +26,107 @@ const IGNORED_TASK_DEFINITION_ATTRIBUTES = [
   'registeredBy'
 ];
 
+// Run task outside of a service
+async function runTask(ecs, clusterName, taskDefArn, waitForMinutes) {
+  core.info('Running task')
+
+  const waitForTask = core.getInput('wait-for-task-stopped', { required: false }) || 'false';
+  const startedBy = core.getInput('run-task-started-by', { required: false }) || 'GitHub-Actions';
+  const launchType = core.getInput('run-task-launch-type', { required: false }) || 'FARGATE';
```
Comment (redterror) [inline dist/index.js:35]: For users with a capacity provider, the `launchType` param is supposed to be omitted.  The implicit default of `FARGATE` here would prevent that.

**Comment by boserup** [comment]
Any progress on this @inhumantsar? I really think this would be a valuable addition to the action.

**Comment by inhumantsar** [comment]
> Any progress on this @inhumantsar? I really think this would be a valuable addition to the action.

I wouldn't count on this PR being merged anytime soon. No one from AWS seems interested in even looking at it. Even after escalating it to our TAM. 

If I'm being totally honest, the scheduled and ad-hoc task operations seem like they haven't had any love from the ECS team in general, so I'm not optimistic about this execution model's future. I've been recommending that our devs leverage Lambda or Lambda+EC2 for anything like this going forward.

**Comment by muzfuz** [comment]
@inhumantsar thanks for your work on this in any case.

Quite frustrating that the AWS team won't support this package, as it's a very common use case.

**Comment by boserup** [comment]
We ended up using the AWS CLI to run one-off commands after deployments from. 
The trick is to enable `wait-for-service-stability` when deploying the new task and afterwards use the `aws ecs list-tasks` command to determine the ID of the newly created task. 

Then one can simply run a command as such:
```bash
aws ecs execute-command --cluster CLUSTER_NAME --task ${{ steps.ecs_task_id.outputs.TASK_ID }} --container CONTAINER_NAME --interactive --command "uname -a"
```

It is not as clean as I'd like, but it gets the job done in our case.

**Comment by amazreech** [comment]
Hi @felipem1210,

Thank you for your contribution, apologies on the delay. We will be working to review the changes in the Pull Request. In the meantime please ensure that below steps, if not already completed, are taken care of in your Pull Request:

1. Verify if PR follows semantic pull request [conventions](https://www.conventionalcommits.org/en/v1.0.0/#specification).
Example: Update the PR title to follow conventions `feat: add ad-hoc task runs ...`

2. Resolve any conflicts on the PR

3. Please run `npm run package` command to update dist/ folder with latest dependencies.

File: index.test.js [hunk 2]
```diff
@@ -146,6 +150,58 @@ describe('Deploy to ECS', () => {
                 }
             };
         });
+
+        mockRunTask.mockImplementation(() => {
+            return {
+                promise() {
```
Comment (trivikr) [inline index.test.js:156]: The mock needs to be updated as the action is now using JS SDK v3

Refer: https://github.com/aws-actions/amazon-ecs-deploy-task-definition/pull/529

File: index.js [hunk 3]
```diff
@@ -20,6 +20,103 @@ const IGNORED_TASK_DEFINITION_ATTRIBUTES = [
...
+  const runTaskResponse = await ecs.runTask({
+    startedBy: startedBy,
+    cluster: clusterName,
+    taskDefinition: taskDefArn,
+    overrides: {
+      containerOverrides: containerOverrides
+    },
+    launchType: launchType,
+    networkConfiguration: awsvpcConfiguration === {} ? {} : { awsvpcConfiguration: awsvpcConfiguration }
+  }).promise();
```
Comment (trivikr) [inline index.js:52]:
```suggestion
  });
```

Comment (trivikr) [inline index.js:52]:
```suggestion
  const runTaskResponse = await ecs.runTask({
    startedBy: startedBy,
    cluster: clusterName,
    taskDefinition: taskDefArn,
    overrides: {
      containerOverrides: containerOverrides
    },
    launchType: launchType,
    networkConfiguration: awsvpcConfiguration === {} ? {} : { awsvpcConfiguration: awsvpcConfiguration }
  });
```

File: index.js [hunk 4]
```diff
@@ -20,6 +20,103 @@ const IGNORED_TASK_DEFINITION_ATTRIBUTES = [
...
+  core.info(`Waiting for tasks to stop. Will wait for ${waitForMinutes} minutes`);
+
+  const waitTaskResponse = await ecs.waitFor('tasksStopped', {
+    cluster: clusterName,
+    tasks: taskArns,
+    $waiter: {
+      delay: WAIT_DEFAULT_DELAY_SEC,
+      maxAttempts: (waitForMinutes * 60) / WAIT_DEFAULT_DELAY_SEC
+    }
+  }).promise();
```
Comment (trivikr) [inline index.js:89]:
```suggestion
  const waitTaskResponse = await waitUntilTasksStopped({
    client: ecs,
    minDelay: WAIT_DEFAULT_DELAY_SEC,
    maxWaitTime: waitForMinutes * 60,
  }, {
    cluster: clusterName,
    tasks: taskArns,
  });
```

File: index.js [hunk 5]
```diff
@@ -20,6 +20,103 @@ const IGNORED_TASK_DEFINITION_ATTRIBUTES = [
...
+  core.debug(`Run task response ${JSON.stringify(waitTaskResponse)}`);
+  core.info('All tasks have stopped.');
+}
+
+// Check a task's exit code and fail the job on error
+async function tasksExitCode(ecs, clusterName, taskArns) {
+  const describeResponse = await ecs.describeTasks({
+    cluster: clusterName,
+    tasks: taskArns
+  }).promise();
```
Comment (trivikr) [inline index.js:100]:
```suggestion
  const describeResponse = await ecs.describeTasks({
    cluster: clusterName,
    tasks: taskArns
  });
```

**Review by trivikr — COMMENTED** [review]
Provided diff for transforming JS SDK v2 APIs to v3.

Refs: https://github.com/aws-actions/amazon-ecs-deploy-task-definition/pull/529

**Comment by inhumantsar** [comment]
thanks @trivikr! i'll commit these and rebase later today

**Comment by inhumantsar** [comment]
alright, i've dealt with the conflicts and updated the tests. should be good to go. @amazreech

**Review by omkhegde — APPROVED** [review]

**Review by kg-aws — APPROVED** [review]

**Comment by inhumantsar** [comment]
🎉 thanks!

**Pull request merged**